[OpenXR][GTK][WPE] Don't request exit XR session before a session started

fujii opened the pull request:
#### 1dbd421437f8d1929d74ae8bb8381c9e23b64702
<pre>
[OpenXR][GTK][WPE] Don&apos;t request exit XR session before a session started
<a href="https://bugs.webkit.org/show_bug.cgi?id=299082">https://bugs.webkit.org/show_bug.cgi?id=299082</a>

Reviewed by Adrian Perez de Castro.

Calling xrRequestExitSession() was failing if an XR session was closed
immediately after the session was created. xrBeginSession() wasn&apos;t
called yet at the time. Check m_isSessionRunning before calling
xrRequestExitSession().

If only with the above change, UIClient::didEndXRSession() would be
called without UIClient::didStartXRSession() was called. We have two
choices to solve it.

1. Calling UIClient::didEndXRSession() when the state becomes XR_SESSION_STATE_STOPPING.
   UIClient::didStartXRSession() was called on XR_SESSION_STATE_READY.

2. Calling UIClient::didStartXRSession() in OpenXRCoordinator::startSession.
   UIClient::didEndXRSession() is called in OpenXRCoordinator::endSessionIfExists.

This change takes 2.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::handleSessionStateChange):

Canonical link: <a href="https://commits.webkit.org/300602@main">https://commits.webkit.org/300602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79ae15246462dcfc8f4487cf091a793f80524f28

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121709 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41410 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32074 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128236 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73799 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92478 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61488 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109005 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73139 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32614 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71752 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131024 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48644 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36997 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101048 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49016 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105222 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100940 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24423 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45345 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19410 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48502 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54229 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47971 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51321 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49654 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/79ae15246462dcfc8f4487cf091a793f80524f28. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50977

fujii commented:
> Looks good but we definitely need an API test that verifies that this fixes the bug and that we don't regress.

I found out this bug while creating a API test for `WebKitXRPermissionRequest`.
I will add a test case later. The test will cover this case. Without this fix, I can't add the test case.
Could you approve this change without a test at the moment?

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
Okay, let's land this as-is for now, and open a new issue to improve the error handling later.

fujii committed:
[OpenXR][GTK][WPE] Don't request exit XR session before a session started
https://bugs.webkit.org/show_bug.cgi?id=299082

Reviewed by Adrian Perez de Castro.

Calling xrRequestExitSession() was failing if an XR session was closed
immediately after the session was created. xrBeginSession() wasn't
called yet at the time. Check m_isSessionRunning before calling
xrRequestExitSession().

If only with the above change, UIClient::didEndXRSession() would be
called without UIClient::didStartXRSession() was called. We have two
choices to solve it.

1. Calling UIClient::didEndXRSession() when the state becomes XR_SESSION_STATE_STOPPING.
   UIClient::didStartXRSession() was called on XR_SESSION_STATE_READY.

2. Calling UIClient::didStartXRSession() in OpenXRCoordinator::startSession.
   UIClient::didEndXRSession() is called in OpenXRCoordinator::endSessionIfExists.

This change takes 2.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::handleSessionStateChange):

Canonical link: https://commits.webkit.org/300602@main

webkit-commit-queue commented:
Committed 300602@main (1dbd421437f8): <https://commits.webkit.org/300602@main>

Reviewed commits have been landed. Closing PR #50977 and removing active labels.
